{"cells": [{"metadata": {"id": "d6db22e49d9a48109172eb912e22151f"}, "cell_type": "markdown", "source": "# notebook1-ncbi-ftp"}, {"metadata": {"id": "7da971eca48f4b05b10099e7932cef12"}, "cell_type": "raw", "source": "!pip install wget"}, {"metadata": {"id": "57772b113c914fc8847edcebfd21b28b"}, "cell_type": "raw", "source": "wslib.download_file(\"contacts_fjgreco_us_ibm_com.csv\")"}, {"metadata": {"id": "9bb9790dfd86453b87950d2291447aa1"}, "cell_type": "raw", "source": "help(wslib.upload_file)"}, {"metadata": {"id": "ec2cde9d8732407d8f3c94caf6215c56"}, "cell_type": "raw", "source": "help(wslib.save_data)"}, {"metadata": {"id": "1347f54e8ecd48d7b41399d7126bee6d"}, "cell_type": "raw", "source": "help(wslib.get_connected_data)"}, {"metadata": {"id": "2f8f1cb93ce3480481bc0d54997cad17"}, "cell_type": "code", "source": "from ibm_watson_studio_lib import access_project_or_space\nwslib = access_project_or_space()", "execution_count": 1, "outputs": []}, {"metadata": {"id": "af55e7d9-16e0-40dd-9951-7d177c5fbc98"}, "cell_type": "code", "source": "import pandas as pd\nimport json", "execution_count": 2, "outputs": []}, {"metadata": {"id": "c8fd656a11f8423c864e93c68075f8c8"}, "cell_type": "markdown", "source": "## Review project's connected data assets"}, {"metadata": {"id": "21ceb41fbde6469e868406e09d0a96c3"}, "cell_type": "code", "source": "ncbi_credentials = wslib.get_connection(\"ncbi\")\n\nncbi_credentials", "execution_count": 3, "outputs": [{"data": {"text/plain": "{'port': '21',\n 'host': 'ftp.ncbi.nih.gov',\n 'connection_mode': 'basic',\n 'username': 'anonymous',\n '.': {'name': 'ncbi',\n  'description': '',\n  'asset_id': '0d95880f-3c3e-44d4-bf8d-cd98787f3c38',\n  'asset_type': 'connection',\n  'personal_credentials': True}}"}, "metadata": {}, "execution_count": 3, "output_type": "execute_result"}]}, {"metadata": {"id": "876d8e0b517e4c309ee57753d39ce9eb"}, "cell_type": "code", "source": "wslib.list_connected_data()", "execution_count": 4, "outputs": [{"data": {"text/plain": "[{'name': 'refseq-genbank.csv',\n  'description': 'ncbi : bioproject/refseq-genbank.csv',\n  'asset_id': 'a7bc469f-82f3-466f-91cf-ef8391af1401',\n  'asset_type': 'data_asset',\n  'tags': ['connected-data']}]"}, "metadata": {}, "execution_count": 4, "output_type": "execute_result"}]}, {"metadata": {"id": "98d6afd9a55240cd9f0fbdc8edf4de8e"}, "cell_type": "code", "source": "z=wslib.get_connected_data('refseq-genbank.csv')\n\nprint(json.dumps(z,indent=2))\n\n", "execution_count": 5, "outputs": [{"name": "stdout", "text": "{\n  \"port\": \"21\",\n  \"host\": \"ftp.ncbi.nih.gov\",\n  \"connection_mode\": \"basic\",\n  \"username\": \"anonymous\",\n  \"datapath\": \"bioproject/refseq-genbank.csv\",\n  \".\": {\n    \"name\": \"refseq-genbank.csv\",\n    \"description\": \"ncbi : bioproject/refseq-genbank.csv\",\n    \"asset_id\": \"a7bc469f-82f3-466f-91cf-ef8391af1401\",\n    \"asset_type\": \"data_asset\",\n    \"asset_state\": \"available\",\n    \"tags\": [\n      \"connected-data\"\n    ],\n    \"connection\": {\n      \"name\": \"ncbi\",\n      \"description\": \"\",\n      \"asset_id\": \"0d95880f-3c3e-44d4-bf8d-cd98787f3c38\",\n      \"asset_type\": \"connection\",\n      \"personal_credentials\": true\n    }\n  }\n}\n", "output_type": "stream"}]}, {"metadata": {"id": "2b192b84b03f46b0830d3e25da204847"}, "cell_type": "markdown", "source": "## Pull from FTP site using project asset information"}, {"metadata": {"id": "4ee59a6cbda94fea8516bc2c32d8df05"}, "cell_type": "code", "source": "import ftplib\n\ndef setupFTP(x,cwd='.',listd=True):\n   \n    ftp = ftplib.FTP()\n\n    ftp.connect(x['host'])\n\n    ftp.login(x['username'])\n\n    ftp.cwd(cwd)\n\n    if listd:\n        ftp.retrlines(\"LIST\")\n    return ftp\n\ndef getFTPFile(ftp,source,target,quit=True):\n    localfile = open(target, 'wb')\n    ftp.retrbinary('RETR ' + source, localfile.write, 1024)\n    localfile.close()\n\n    if quit:\n        print(ftp.quit())\n        \ndef getFTPList(ftp,listd=True):\n    data = []\n\n    ftp.dir(data.append)\n\n    if listd:\n        for line in data:\n            print (\"-\", line)\n            \n    return data\n        \n    \ndef ftp_quit(ftp):\n    try:\n        ftp.quit()\n    except:\n        print(\"Already quit\")\n        \ndef downloadConnectedData(source,target):\n    try:\n        x=wslib.get_connected_data(source)\n        \n        print(json.dumps(x,indent=2))\n\n        print(x['host'], x['username'],x['datapath'])\n\n        ftp=setupFTP(x)\n\n        getFTPFile(ftp,x['datapath'],target)\n    except:\n        print('error')\n        ftp_quit(ftp)\n\n    ", "execution_count": 7, "outputs": []}, {"metadata": {"id": "a98e3415bf424a57826321881761b77f"}, "cell_type": "markdown", "source": "## Download  asset from FTP site "}, {"metadata": {"id": "9ea5379f3b334b198a24d58a31957e45"}, "cell_type": "code", "source": "downloadConnectedData(\"refseq-genbank.csv\",\"refseq-genbank_local.csv\")", "execution_count": 8, "outputs": [{"name": "stdout", "text": "{\n  \"port\": \"21\",\n  \"host\": \"ftp.ncbi.nih.gov\",\n  \"connection_mode\": \"basic\",\n  \"username\": \"anonymous\",\n  \"datapath\": \"bioproject/refseq-genbank.csv\",\n  \".\": {\n    \"name\": \"refseq-genbank.csv\",\n    \"description\": \"ncbi : bioproject/refseq-genbank.csv\",\n    \"asset_id\": \"a7bc469f-82f3-466f-91cf-ef8391af1401\",\n    \"asset_type\": \"data_asset\",\n    \"asset_state\": \"available\",\n    \"tags\": [\n      \"connected-data\"\n    ],\n    \"connection\": {\n      \"name\": \"ncbi\",\n      \"description\": \"\",\n      \"asset_id\": \"0d95880f-3c3e-44d4-bf8d-cd98787f3c38\",\n      \"asset_type\": \"connection\",\n      \"personal_credentials\": true\n    }\n  }\n}\nftp.ncbi.nih.gov anonymous bioproject/refseq-genbank.csv\ndr-xr-xr-x   4 ftp      anonymous     4096 Mar 27 02:48 1000genomes\n-r--r--r--   1 ftp      anonymous 10737418240 Sep 18  2019 10GB\n-r--r--r--   1 ftp      anonymous 1073741824 Sep 18  2019 1GB\n-r--r--r--   1 ftp      anonymous 53687091200 Sep 18  2019 50GB\n-r--r--r--   1 ftp      anonymous 5368709120 Sep 18  2019 5GB\n-r--r--r--   1 ftp      anonymous     2037 Sep 13  2019 README.ftp\ndr-xr-xr-x 1073741824 ftp      anonymous       40 Mar  9 13:27 ReferenceSamples\ndr-xr-xr-x   8 ftp      anonymous     4096 Feb  8 03:48 SampleData\nlr--r--r--   1 ftp      anonymous       29 Sep 13  2019 asn1-converters -> toolbox/ncbi_tools/converters\ndr-xr-xr-x  12 ftp      anonymous   188416 Feb  8 03:48 bigwig\ndr-xr-xr-x   5 ftp      anonymous     4096 Mar 27 15:10 bioproject\ndr-xr-xr-x   2 ftp      anonymous     4096 Mar 27 12:05 biosample\ndr-xr-xr-x  13 ftp      anonymous     4096 Mar 27 02:48 blast\ndr-xr-xr-x   3 ftp      anonymous     4096 Sep 13  2004 cgap\ndr-xr-xr-x   4 ftp      anonymous     4096 Oct  3  2014 cn3d\ndr-xr-xr-x   2 ftp      anonymous    28672 Mar 27 02:48 comparative-genome-viewer\ndr-xr-xr-x  38 ftp      anonymous     4096 Feb  8 03:48 dbgap\ndr-xr-xr-x   2 ftp      anonymous     4096 Sep 21  2017 diffexpIR-notebook\ndr-xr-xr-x  12 ftp      anonymous     4096 Jul 18  2013 entrez\ndr-xr-xr-x   7 ftp      anonymous     4096 Feb  8 03:48 epigenomics\ndr-xr-xr-x   4 ftp      anonymous     4096 Sep 19  2017 eqtl\ndr-xr-xr-x   6 ftp      anonymous     4096 Aug  4  2006 fa2htgs\n-r--r--r--   1 ftp      anonymous     3262 Sep 13  2019 favicon.ico\n-r--r--r--   1 ftp      anonymous      106 Feb 16 13:24 fufuter.html\ndr-xr-xr-x  13 ftp      anonymous   376832 Mar 27 02:48 genbank\ndr-xr-xr-x   6 ftp      anonymous     4096 Feb  8 03:48 gene\ndr-xr-xr-x  15 ftp      anonymous    49152 Mar 27 02:48 genomes\ndr-xr-xr-x 1073741824 ftp      anonymous        0 Mar 27 13:21 geo\ndr-xr-xr-x   4 ftp      anonymous     4096 Mar 27 02:48 giab\ndr-xr-xr-x  25 ftp      anonymous     4096 Sep 20  2011 hapmap\ndr-xr-xr-x  12 ftp      anonymous     4096 Mar 21 21:24 hmm\ndr-xr-xr-x  15 ftp      anonymous     4096 Nov 27 21:23 mmdb\ndr-xr-xr-x   7 ftp      anonymous   184320 Mar 27 02:48 ncbi-asn1\ndr-xr-xr-x   2 ftp      anonymous        6 Sep 13  2019 nist-immsa\ndr-xr-xr-x   3 ftp      anonymous    12288 Sep  1  2021 osiris\ndr-xr-xr-x   9 ftp      anonymous     4096 Mar 27 02:48 pathogen\ndr-xr-xr-x 171 ftp      anonymous     8192 Oct 19 23:53 pub\ndr-xr-xr-x  22 ftp      anonymous     8192 Mar  7 14:03 pubchem\ndr-xr-xr-x   3 ftp      anonymous     4096 Apr 19  2013 pubchem-scratch\ndr-xr-xr-x  28 ftp      anonymous     4096 Dec 13 14:48 pubmed\ndr-xr-xr-x   4 ftp      anonymous     4096 Feb  8 03:48 rapt\ndr-xr-xr-x 1073741824 ftp      anonymous       40 Mar  9 13:27 refsam\ndr-xr-xr-x  22 ftp      anonymous     4096 Mar 11 19:30 refseq\ndr-xr-xr-x  57 ftp      anonymous     4096 Feb  8 03:48 repository\n-r--r--r--   1 ftp      anonymous       26 Sep 13  2019 robots.txt\ndr-xr-xr-x   3 ftp      anonymous     4096 Mar 27 02:48 seqc\ndr-xr-xr-x   3 ftp      anonymous     4096 Jan 26  2021 sequin\ndr-xr-xr-x   4 ftp      anonymous     4096 Jun 23  2016 sky-cgh\ndr-xr-xr-x  11 ftp      anonymous     4096 Mar 27 02:48 snp\ndr-xr-xr-x  23 ftp      anonymous     4096 Mar 27 02:48 sra\ndr-xr-xr-x   2 ftp      anonymous     4096 Sep 29  2004 tech-reports\ndr-xr-xr-x  11 ftp      anonymous     4096 Jul 25  2013 toolbox\ndr-xr-xr-x   8 ftp      anonymous     4096 Sep 19  2021 tpa\ndr-xr-xr-x   5 ftp      anonymous     4096 Aug  9  2016 variation\n221 Goodbye.\n", "output_type": "stream"}]}, {"metadata": {"id": "273f0e86b3fe4c259e01be5e02481f64"}, "cell_type": "raw", "source": "ftp_quit(ftp)"}, {"metadata": {"id": "0ca5ad75d1784db981c237c54bbf57b2"}, "cell_type": "markdown", "source": "### List local storage contents"}, {"metadata": {"id": "71c194e4ab25499fbc90d8ca7eed0eae"}, "cell_type": "code", "source": "!ls -al", "execution_count": 17, "outputs": [{"name": "stdout", "text": "total 248\r\ndrwxr-x---. 3 1000690000 wscommon  4096 Mar 27 19:02 .\r\ndrwxrwx---. 1 wsuser     wsbuild   4096 Mar 27 15:51 ..\r\n-rw-rw----. 1 1000690000 wscommon 34231 Mar 27 16:32 contacts_fjgreco_us_ibm_com.csv\r\n-rw-rw----. 1 1000690000 wscommon 93506 Mar 27 17:20 refseq-genbank.csv\r\n-rw-rw----. 1 1000690000 wscommon 93506 Mar 27 19:02 refseq-genbank_local.csv\r\ndrwxr-x---. 2 1000690000 wscommon  4096 Mar 27 15:50 .virtual_documents\r\n", "output_type": "stream"}]}, {"metadata": {"id": "91b8ea2a6e63470f8c68ef54f63f0e37"}, "cell_type": "markdown", "source": "## View recently downloaded file in pandas"}, {"metadata": {"id": "8c7369cf3b8e45fa8ec4387ec4399d22"}, "cell_type": "code", "source": "pd.read_csv(\"refseq-genbank_local.csv\")", "execution_count": 16, "outputs": [{"data": {"text/plain": "      Refseq accn Genbank accn                   Organism name    TaxID\n0        PRJNA116   PRJNA10719            Arabidopsis thaliana     3702\n1        PRJNA116   PRJNA11796            Arabidopsis thaliana     3702\n2        PRJNA116   PRJNA13191            Arabidopsis thaliana     3702\n3        PRJNA122   PRJNA12269     Oryza sativa Japonica Group    39947\n4        PRJNA122    PRJDB1747     Oryza sativa Japonica Group    39947\n...           ...          ...                             ...      ...\n1714  PRJNA756971  PRJNA682572        Prionailurus bengalensis    37029\n1715  PRJNA758027    PRJDB3949            Aspergillus udagawae    91492\n1716  PRJNA758049    PRJDB7449  Aspergillus pseudoviridinutans  1517512\n1717  PRJNA759178  PRJNA597580                  Colletes gigas   935657\n1718  PRJNA759270  PRJNA736740              Venturia canescens    32260\n\n[1719 rows x 4 columns]", "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>Refseq accn</th>\n      <th>Genbank accn</th>\n      <th>Organism name</th>\n      <th>TaxID</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>PRJNA116</td>\n      <td>PRJNA10719</td>\n      <td>Arabidopsis thaliana</td>\n      <td>3702</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>PRJNA116</td>\n      <td>PRJNA11796</td>\n      <td>Arabidopsis thaliana</td>\n      <td>3702</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>PRJNA116</td>\n      <td>PRJNA13191</td>\n      <td>Arabidopsis thaliana</td>\n      <td>3702</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>PRJNA122</td>\n      <td>PRJNA12269</td>\n      <td>Oryza sativa Japonica Group</td>\n      <td>39947</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>PRJNA122</td>\n      <td>PRJDB1747</td>\n      <td>Oryza sativa Japonica Group</td>\n      <td>39947</td>\n    </tr>\n    <tr>\n      <th>...</th>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n    </tr>\n    <tr>\n      <th>1714</th>\n      <td>PRJNA756971</td>\n      <td>PRJNA682572</td>\n      <td>Prionailurus bengalensis</td>\n      <td>37029</td>\n    </tr>\n    <tr>\n      <th>1715</th>\n      <td>PRJNA758027</td>\n      <td>PRJDB3949</td>\n      <td>Aspergillus udagawae</td>\n      <td>91492</td>\n    </tr>\n    <tr>\n      <th>1716</th>\n      <td>PRJNA758049</td>\n      <td>PRJDB7449</td>\n      <td>Aspergillus pseudoviridinutans</td>\n      <td>1517512</td>\n    </tr>\n    <tr>\n      <th>1717</th>\n      <td>PRJNA759178</td>\n      <td>PRJNA597580</td>\n      <td>Colletes gigas</td>\n      <td>935657</td>\n    </tr>\n    <tr>\n      <th>1718</th>\n      <td>PRJNA759270</td>\n      <td>PRJNA736740</td>\n      <td>Venturia canescens</td>\n      <td>32260</td>\n    </tr>\n  </tbody>\n</table>\n<p>1719 rows \u00d7 4 columns</p>\n</div>"}, "metadata": {}, "execution_count": 16, "output_type": "execute_result"}]}, {"metadata": {"id": "ab5373cc6f2f41c5939abbaf33a4666d"}, "cell_type": "raw", "source": "import itc_utils.flight_service as itcfs\n\nreadClient = itcfs.get_flight_client()\n\nnb_data_request = {\n    'data_name': \"\"\"refseq-genbank_local.csv\"\"\",\n    'interaction_properties': {\n        #'row_limit': 500\n    }\n}\n\nflightInfo = itcfs.get_flight_info(readClient, nb_data_request=nb_data_request)\n\ndata_df_1 = itcfs.read_pandas_and_concat(readClient, flightInfo)\ndata_df_1.head(10)\n"}, {"metadata": {"id": "5ee4b6158fa24abd9459c39433727666"}, "cell_type": "markdown", "source": "### List Project locally stored data assets"}, {"metadata": {"id": "ee74d4c3e28e464bbec4ed31c0c92768"}, "cell_type": "code", "source": "!ls /project_data/data_asset", "execution_count": 10, "outputs": [{"name": "stdout", "text": "contacts_fjgreco_us_ibm_com.csv\r\n", "output_type": "stream"}]}, {"metadata": {"id": "1bf643ce73524c80963c0e44150c9276"}, "cell_type": "markdown", "source": "### List project data assets including connected assets"}, {"metadata": {"id": "a8ec9ab08a6f448e957da23dc9307adb"}, "cell_type": "code", "source": "wslib.assets.list_assets('data_asset')", "execution_count": 11, "outputs": [{"data": {"text/plain": "[{'name': 'contacts_fjgreco_us_ibm_com.csv',\n  'description': '',\n  'asset_id': 'fa0b7995-55bf-4530-9cc6-6b24f8837848',\n  'asset_type': 'data_asset',\n  'tags': []},\n {'name': 'refseq-genbank.csv',\n  'description': 'ncbi : bioproject/refseq-genbank.csv',\n  'asset_id': 'a7bc469f-82f3-466f-91cf-ef8391af1401',\n  'asset_type': 'data_asset',\n  'tags': ['connected-data']}]"}, "metadata": {}, "execution_count": 11, "output_type": "execute_result"}]}, {"metadata": {"id": "e43549fca36442f2988b515823546672"}, "cell_type": "markdown", "source": "### Upload locally saved file to project asset storage"}, {"metadata": {"id": "b6029acf6b72497c89b630c16323d07e"}, "cell_type": "code", "source": "wslib.upload_file(\"refseq-genbank_local.csv\")", "execution_count": 12, "outputs": [{"data": {"text/plain": "{'name': 'refseq-genbank_local.csv',\n 'asset_type': 'data_asset',\n 'asset_id': '8a0b7b12-e040-4fb0-bf5a-d08c0b2e458f',\n 'attachment_id': '3482b97b-3e83-43d5-8ad4-29bae3fc1814',\n 'filepath': 'refseq-genbank_local.csv',\n 'data_size': None,\n 'mime': 'text/csv',\n 'summary': ['created file in storage', 'created asset', 'created attachment'],\n 'input_file_copied': True}"}, "metadata": {}, "execution_count": 12, "output_type": "execute_result"}]}, {"metadata": {"id": "a0d8a4a1b7414cd78e7c0c62119c6c25"}, "cell_type": "markdown", "source": "### List project data assets including connected assets and recently uploaded assets"}, {"metadata": {"id": "47bdf48515e745b0a9f5cfe8777c3429"}, "cell_type": "code", "source": "wslib.assets.list_assets('data_asset')", "execution_count": 14, "outputs": [{"data": {"text/plain": "[{'name': 'refseq-genbank.csv',\n  'description': 'ncbi : bioproject/refseq-genbank.csv',\n  'asset_id': 'a7bc469f-82f3-466f-91cf-ef8391af1401',\n  'asset_type': 'data_asset',\n  'tags': ['connected-data']},\n {'name': 'refseq-genbank_local.csv',\n  'description': '',\n  'asset_id': '8a0b7b12-e040-4fb0-bf5a-d08c0b2e458f',\n  'asset_type': 'data_asset',\n  'tags': []},\n {'name': 'contacts_fjgreco_us_ibm_com.csv',\n  'description': '',\n  'asset_id': 'fa0b7995-55bf-4530-9cc6-6b24f8837848',\n  'asset_type': 'data_asset',\n  'tags': []}]"}, "metadata": {}, "execution_count": 14, "output_type": "execute_result"}]}, {"metadata": {"id": "66ec862d3377427c8afcc15da04c6af1"}, "cell_type": "code", "source": "", "execution_count": null, "outputs": []}, {"metadata": {"id": "dbd71815387f49a699886102319f3354"}, "cell_type": "code", "source": "", "execution_count": null, "outputs": []}], "metadata": {"kernelspec": {"name": "python3", "display_name": "Python 3.9", "language": "python"}, "language_info": {"name": "python", "version": "3.9.7", "mimetype": "text/x-python", "codemirror_mode": {"name": "ipython", "version": 3}, "pygments_lexer": "ipython3", "nbconvert_exporter": "python", "file_extension": ".py"}}, "nbformat": 4, "nbformat_minor": 1}